File: packages/browser/ios/DefaultBrowserLauncherAdapter.swift
//
//  DefaultBrowserLauncherAdapter.swift
//  RNPingBrowser
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import PingBrowser
import PingLogger

/// Default adapter that delegates to the Ping BrowserLauncher singleton.
public struct DefaultBrowserLauncherAdapter: BrowserLaunching {
  public init() {}

  // TODO: Accept a `logger:` parameter so the JS-resolved logger can be
  // forwarded to BrowserLauncher.launch() instead of the global LogManager
  // singleton. Requires extending the BrowserLaunching protocol too.
  public func launch(
    url: URL,
    customParams: [String: String]?,
    browserType: BrowserType,
    browserMode: BrowserMode,
    callbackURLScheme: String
  ) async throws -> URL {
    return try await BrowserLauncher.currentBrowser.launch(
      url: url,
      customParams: customParams,
      browserType: browserType,
      browserMode: browserMode,
      callbackURLScheme: callbackURLScheme,
      logger: LogManager.logger
    )
  }

  public func reset() {
    BrowserLauncher.currentBrowser.reset()
  }
}

File: packages/browser/android/src/main/java/com/pingidentity/reactnative/rnbrowser/DefaultBrowserLauncherAdapter.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.reactnative.rnbrowser

import android.content.Intent
import android.net.Uri
import androidx.browser.auth.AuthTabIntent
import androidx.browser.customtabs.CustomTabsIntent
import com.pingidentity.browser.BrowserLauncher
import java.net.URL

/**
 * Default adapter that delegates to the Ping BrowserLauncher singleton.
 */
internal object DefaultBrowserLauncherAdapter : BrowserLauncherAdapter {
  override var customTabsCustomizer: CustomTabsIntent.Builder.() -> Unit
    get() = BrowserLauncher.customTabsCustomizer
    set(value) {
      BrowserLauncher.customTabsCustomizer = value
    }

  override var authTabCustomizer: AuthTabIntent.Builder.() -> Unit
    get() = BrowserLauncher.authTabCustomizer
    set(value) {
      BrowserLauncher.authTabCustomizer = value
    }

  override var intentCustomizer: Intent.() -> Unit
    get() = BrowserLauncher.intentCustomizer
    set(value) {
      BrowserLauncher.intentCustomizer = value
    }

  override var redirectUri: Uri
    get() = BrowserLauncher.redirectUri
    set(value) {
      BrowserLauncher.redirectUri = value
    }

  override suspend fun launch(url: URL, redirectUri: Uri): Result<Uri> {
    return BrowserLauncher.launch(url, redirectUri)
  }
}
